### ISSUE TITLE
LED Matrix Capabilities?

### ISSUE THREAD

**Issue by hendursaga** [espressoshock/ble-ledly#7]
Are you open to supporting LED "matrices" and not just strips? I have one BLE-enabled LED 'badge' that I got most of the protocol worked out, I just wasn't sure if that would be out-of-scope for this project.

**Comment by espressoshock** [comment]
Yes, that's totally within the scope of this library. This library has been designed to be _easily extended_, so it should be pretty straightforward to implement LED matrix-based devices. It can be easily achieved by simply: 

1. implementing your own communication protocol using the `Protocol` trait
2. creating a new capability (i.e. `Matrix`), with high-level functionalities

If you provide more details we can easily add this capability to the crate so more people can benefit from

**Comment by hendursaga** [comment]
I'm relatively new to Rust - I've only created a game in Bevy, so far - so perhaps implementing the `Brightness` trait for the new protocol would be a start? I'm not sure what to call the protocol though - my device is some generic Chinese badge similar to https://www.amazon.com/VBLab-Upgraded-Wireless-Bluetooth-Rechargeable/dp/B07P5X28JY - the app I've used to control it is called [CoolLED1248](https://play.google.com/store/apps/details?id=com.jtkj.led1248&gl=US). I was hoping the open-source [Badge Magic](https://f-droid.org/en/packages/org.fossasia.badgemagic/) would work with it, as they are really similar, but alas, all signs point to a [white-label product](https://en.wikipedia.org/wiki/White-label_product), although if you know more about tracking down the actual company, that would be cool!

**Comment by espressoshock** [comment]
To achieve what you're looking for and control the badge and possibly implement all the features that are currently supported by the app you linked, as mentioned before,  you'd need to implement the `communication protocol` used by the device. 

This should be pretty trivial as by running a quick search people already reversed-engineered it ([here](http://nilhcem.com/iot/reverse-engineering-bluetooth-led-name-badge) and [here](https://cat-in-136.github.io/2020/07/b1248-led-name-badge-protocol-reverse-engineering.html) ).

I'll look into it next week and attach a PR with some basic functionality to drive the badge, but you'd need to test it since I don't have this device on hand.

**Comment by hendursaga** [comment]
Unfortunately even a more extensive search showed no one who specifically RE'd my model of badge, but some of them get rather close, thankfully. I could certainly help with narrowing things down once a more general framework is in place!
